linera-core/src/unit_tests/worker_tests.rs

```diff
@@ -3663,3 +3666,204 @@ where
     assert_eq!(manager.leader, Some(Owner::from(validator_key)));
     Ok(())
 }
+
+/// Tests if a service is able to handle more than one query without restarting.
+///
+/// If the service is restarted, a new [`MockApplicationInstance`] is created with an empty list of
+/// expected calls, and the test fails because the first [`MockApplicationInstance`] still expects
+/// some calls.
+#[test_case(MemoryStorageBuilder::default(); "memory")]
+#[cfg_attr(feature = "rocksdb", test_case(RocksDbStorageBuilder::new().await; "rocks_db"))]
+#[cfg_attr(feature = "aws", test_case(DynamoDbStorageBuilder::default(); "dynamo_db"))]
+#[cfg_attr(feature = "scylladb", test_case(ScyllaDbStorageBuilder::default(); "scylla_db"))]
+#[test_log::test(tokio::test(start_paused = true))]
+async fn test_long_lived_service<B>(mut storage_builder: B) -> anyhow::Result<()>
+where
+    B: StorageBuilder,
+    ViewError: From<<B::Storage as Storage>::ContextError>,
+{
+    const NUM_QUERIES: usize = 5;
+
+    let storage = storage_builder.build().await?;
+    let clock = storage_builder.clock();
+    let chain_description = ChainDescription::Root(1);
+    let chain_id = ChainId::from(chain_description);
+    let key_pair = KeyPair::generate();
+    let balance = Amount::ZERO;
+
+    let (_committee, mut worker) = init_worker_with_chain(
+        storage.clone(),
+        chain_description,
+        key_pair.public(),
+        balance,
+    )
+    .await;
+
+    let mut applications;
+    {
+        let mut chain = storage.load_chain(chain_id).await?;
+        applications = register_mock_applications(&mut chain.execution_state, 1).await?;
+        chain.save().await?;
+    }
+
+    let (application_id, application) = applications
+        .next()
+        .expect("Mock application should be registered");
+
+    let query_context = QueryContext {
+        chain_id,
+        next_block_height: BlockHeight(0),
+        local_time: Timestamp::from(0),
+    };
+    for query_time in 0..NUM_QUERIES {
+        let query_context = QueryContext {
+            local_time: Timestamp::from(query_time as u64),
+            ..query_context
+        };
+
+        application.expect_call(ExpectedCall::handle_query(
+            move |_runtime, context, query| {
+                assert_eq!(context, query_context);
+                assert!(query.is_empty());
+                Ok(vec![])
+            },
+        ));
+    }
+
+    let query = Query::User {
+        application_id,
+        bytes: vec![],
+    };
+    for query_time in 0..NUM_QUERIES {
+        clock.set(Timestamp::from(query_time as u64));
+
+        assert_eq!(
+            worker.query_application(chain_id, query.clone()).await?,
+            Response::User(vec![])
+        );
+    }
+
+    drop(worker);
+    tokio::time::sleep(Duration::from_secs(1)).await;
+    application.assert_no_more_expected_calls();
+
+    Ok(())
+}
+
+/// Tests if a service is restarted when a block is added to the chain.
```

Comment:
But it's only proposed, not confirmed, is it? So it's not really added to the chain. Otherwise the block height in the query context would change, too.

You could add a third sequence of queries, after the confirmed block certificate was handled by the worker, to verify that.

Reply:
Thanks for the tip! Done.

Reply:
(And the fact that it restarts whenever a block is _proposed_ could be considered a shortcoming rather than a feature. But I guess it's fine to test the currently expected behavior anyway.)